Source/Core/VideoBackends/Vulkan/VKShader.cpp

```diff
@@ -0,0 +1,106 @@
+// Copyright 2017 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include <algorithm>
+#include <cstddef>
+
+#include "Common/Assert.h"
+#include "Common/CommonTypes.h"
+#include "Common/Logging/Log.h"
+
+#include "VideoBackends/Vulkan/Renderer.h"
+#include "VideoBackends/Vulkan/ShaderCompiler.h"
+#include "VideoBackends/Vulkan/Util.h"
+#include "VideoBackends/Vulkan/VKShader.h"
+#include "VideoBackends/Vulkan/VulkanContext.h"
+
+namespace Vulkan
+{
+VKShader::VKShader(ShaderStage stage, std::vector<u32> spv, VkShaderModule mod)
+    : AbstractShader(stage), m_spv(std::move(spv)), m_module(mod)
+{
+}
+
+VKShader::~VKShader()
+{
+  vkDestroyShaderModule(g_vulkan_context->GetDevice(), m_module, nullptr);
+}
+
+VkShaderModule VKShader::GetShaderModule() const
+{
+  return m_module;
+}
+
+bool VKShader::HasBinary() const
+{
+  // m_spv should always be non-empty.
+  return true;
+}
+
+AbstractShader::BinaryData VKShader::GetBinary() const
+{
+  BinaryData ret;
+  ret.resize(sizeof(u32) * m_spv.size());
+  std::memcpy(ret.data(), m_spv.data(), sizeof(u32) * m_spv.size());
+  return ret;
+}
+
+std::unique_ptr<AbstractShader> Renderer::CreateShaderFromSource(ShaderStage stage,
+                                                                 const char* source, size_t length)
+{
+  ShaderCompiler::SPIRVCodeVector spv;
+  bool result;
+  switch (stage)
+  {
+  case ShaderStage::Vertex:
+    result = ShaderCompiler::CompileVertexShader(&spv, source, length);
+    break;
+  case ShaderStage::Geometry:
+    result = ShaderCompiler::CompileGeometryShader(&spv, source, length);
+    break;
+  case ShaderStage::Pixel:
+    result = ShaderCompiler::CompileFragmentShader(&spv, source, length);
+    break;
+  case ShaderStage::Compute:
+    result = ShaderCompiler::CompileComputeShader(&spv, source, length);
+    break;
+  default:
+    result = false;
+    break;
+  }
+
+  if (!result)
+    return nullptr;
+
+  VkShaderModule mod = Util::CreateShaderModule(spv.data(), spv.size());
+  if (mod == VK_NULL_HANDLE)
+    return nullptr;
+
+  return std::make_unique<VKShader>(stage, std::move(spv), mod);
+}
+
+std::unique_ptr<AbstractShader> Renderer::CreateProgramFromBinary(ShaderStage stage,
+                                                                  const void* data, size_t length)
+{
+  ShaderCompiler::SPIRVCodeVector spv;
+  const size_t size_in_words = sizeof(length) / sizeof(ShaderCompiler::SPIRVCodeType);
+  spv.resize(length / size_in_words);
+  if (size_in_words > 0)
```

Comment:
Is this needed?  If size-in-words is 0, `spv.resize` above will cause divide by zero error..